### PR TITLE
Parse namespaces/types aliases/interfaces/ambient declarations only when they cannot be legal JS

### LINKED ISSUE #2995: Legal JS is interpreted incorrectly in the presence of certain keywords
Take the following code.

```TypeScript
var type;
var string;
var Foo;

type Foo = string;
```

Now add a newline between `type Foo`:

```TypeScript
var type;
var string;
var Foo;

type
Foo = string;
```

Still valid, but should have a completely different meaning now. In the first, we have a type alias, while in the second, we have a lone identifier and a declaration.

-----------------------

Here's actually a few contexts where we do the wrong thing:

* type aliases (see the example above)
* namespaces/internal modules:
```Typescript
namespace    // this is the identifier 'namespace'
m            // this is the identifier 'm'
{ }          // this is a block body
```
* declared modules
```TypeScript
declare                // this is the identifier 'declare'
module                 // this is the identifier 'module'
"my external module"   // this is just a string
{ }                    // this is a block body
```

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #2995.

Previously, upon encountering a `module`, `namespace`, `type`, `interface`, or `declare` keyword, we would immediately launch into parsing out an "appropriate" declaration.

However, this is inappropriate for reasons listed in issue #2995. These could actually be identifiers. To deal with this, we ensure that what we have is not legal JavaScript by adding a single token lookahead and checking for whether that token is an appropriate token *on the same line*.

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -3981,19 +3981,31 @@ module ts {
                     return parseDebuggerStatement();
                 case SyntaxKind.AtToken:
                     return parseDeclaration();
-                case SyntaxKind.ConstKeyword:
                 case SyntaxKind.DeclareKeyword:
+                    if (lookAhead(isFollowedByIdentifierOrKeywordOnSameLine) && getDeclarationFlags() & flags) {
```
Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: You need some comment explaining why the lookahead is necessary here, but not for the syntaxkind kinds below.

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -4044,6 +4056,21 @@ module ts {
             }
         }
 
+        function isFollowedByIdentifierOrKeywordOnSameLine() {
```
Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: We already have: nextTokenIsIdentifierOrKeywordOnSameLine

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -4044,6 +4056,21 @@ module ts {
             }
         }
 
+        function isFollowedByIdentifierOrKeywordOnSameLine() {
+            nextToken();
+            return !scanner.hasPrecedingLineBreak() && isIdentifierOrKeyword();
+        }
+
+        function isFollowedByIdentifierOrStringOnSameLine() {
```
Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: nextTokenIsIdentifierOrStringLiteralOnSameLine

File: src/compiler/parser.ts [hunk 4]
```diff
@@ -4044,6 +4056,21 @@ module ts {
...
+            nextToken();
+            return !scanner.hasPrecedingLineBreak() && isIdentifierOrKeyword();
+        }
+
+        function isFollowedByIdentifierOrStringOnSameLine() {
+            nextToken();
+            return !scanner.hasPrecedingLineBreak() && (isIdentifier() || token === SyntaxKind.StringLiteral);
+        }
+
+        function isFollowedByIdentifierOnSameLine() {
```
Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: nextTokenIsIdentifierOnSameLine

**Comment by DanielRosenwasser** [comment]
Any more feedback @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
Nope.  lgtm

File: src/compiler/parser.ts [hunk 5]
```diff
@@ -3981,23 +3981,57 @@ module ts {
...
+                    break;
+                case SyntaxKind.ModuleKeyword:
+                    if (lookAhead(nextTokenIsIdentifierOrStringLiteralOnSameLine) && getDeclarationFlags() & flags) {
+                        return parseDeclaration();
+                    }
+                    break;
+                case SyntaxKind.InterfaceKeyword:
+                case SyntaxKind.NamespaceKeyword:
+                case SyntaxKind.TypeKeyword:
+                    if (lookAhead(nextTokenIsIdentifierOnSameLine) && getDeclarationFlags() & flags) {
```
Comment (JsonFreeman) [inline src/compiler/parser.ts]: Do you still need these declaration flags checks? Are they redundant of the check you are adding?

**Comment by JsonFreeman** [comment]
:+1:

**Pull request merged**